### PR TITLE
Ensemble grammar, typo, re-wording

### PR DESCRIPTION
**Pull request by lucyleeow**
I think we could expand/give more intuition on the need for the 'base' ensemble estimators to be 'weak' (/not ideal) learners (for both bagging and boosting) and how this avoids overfiting. Maybe e.g., combining weak learners that each focus, not too strongly, on one area/relationship is better than using a 'stronger' learner which is likely to overfit (as it is allowed to 'focus' too much). 
And how/why the bootstrap samples result in 'weak' learners.

### CODE REVIEW

File: python_scripts/ensemble.py [hunk 1]
```diff
@@ -829,13 +840,13 @@ def plot_decision_function(X, y, clf, sample_weight=None, fit=True, ax=None):
 print(f"Error of the tree: {y_true - y_pred_first_and_second_tree:.3f}")
 
 # %% [markdown]
-# We choose a sample for which only two trees were enough to make the perfect
+# We chose a sample for which only two trees were enough to make the perfect
 # prediction. However, we saw in the previous plot that two trees were not
-# enough to correct the residuals for all samples. In this regard, one need to
-# add several trees in the ensemble to succeed to correct the error.
+# enough to correct the residuals of all samples. Therefore, one needs to
+# add several trees to the ensemble to successfully correct the error.
```
Comment (lucyleeow) [inline python_scripts/ensemble.py:846]: Maybe clarify that we are successively fitting the residuals of the previous estimator, i.e., clf1 fits the residuals of clf0, clf2 fits the residuals of clf1 ... Maybe this is obvious from the text, I'm unsure about this nitpick.

File: python_scripts/ensemble.py [hunk 2]
```diff
@@ -907,35 +918,36 @@ def plot_decision_function(X, y, clf, sample_weight=None, fit=True, ax=None):
...
-# largest `n_estimators` lead to the best performance.
+# largest `n_estimators` led to the best performance.
 #
-# Gradient-boosting decision tree
+# ### Gradient-boosting decision tree
 #
-# In gradient-boosting, parameters tuning is a combination of several
+# For gradient-boosting, parameter tuning is a combination of several
 # parameters instead of setting one after the other each parameter. The
 # important parameters are `n_estimators`, `max_depth`, and `learning_rate`.
```
Comment (lucyleeow) [inline python_scripts/ensemble.py:927]: This is not so clear. Have we tuned parameters one at a time previously? We tuned for a combination of parameters?

File: python_scripts/ensemble.py [hunk 3]
```diff
@@ -907,35 +918,36 @@ def plot_decision_function(X, y, clf, sample_weight=None, fit=True, ax=None):
...
-# levels typically.
+# the first tree of the ensemble would perfectly fit (overfit) the data and
+# thus no subsequent tree would be required, since there would be no residuals.
+# Therefore, the tree used in gradient-boosting should have a low depth,
+# typically between 3 to 8 levels.
 #
-# Having in mind this consideration, deeper will be the trees, faster the
+# With this consideration in mind, the deeper the trees, the faster the
 # residuals will be corrected and less learners are required. So `n_estimators`
 # should be increased if `max_depth` is lower.
```
Comment (lucyleeow) [inline python_scripts/ensemble.py:939]: This brings the question when/why would you want deeper trees? For what kind of problems would it be better to have deeper trees and less learners? Or this is general trial and error, see what gives you the best results?

File: python_scripts/ensemble.py [hunk 4]
```diff
@@ -907,35 +918,36 @@ def plot_decision_function(X, y, clf, sample_weight=None, fit=True, ax=None):
...
-# Let's first discuss about the `max_depth` parameter. We saw in the section of
-# the gradient-boosting that the algorithm fit the error of the previous trees
+# Let's first discuss the `max_depth` parameter. We saw in the section on
+# gradient-boosting that the algorithm fits the error of the previous tree
 # in the ensemble. Thus, fitting fully grown trees will be detrimental. Indeed,
-# the first tree of the ensemble will overfit the data and thus no subsequent
-# tree is required since there will be no residuals to be fitted. Therefore,
-# the tree used in gradient-boosting will have a low depth between 3 to 8
-# levels typically.
+# the first tree of the ensemble would perfectly fit (overfit) the data and
```
Comment (lucyleeow) [inline python_scripts/ensemble.py:932]: Maybe expand on this - specifically that the ultimate reason is to avoid overfitting, not just to have residuals to fit. We want weak learners, that sequentially 'weakly' learn the residuals because it avoids over-fitting

File: python_scripts/ensemble.py [hunk 5]
```diff
@@ -907,35 +918,36 @@ def plot_decision_function(X, y, clf, sample_weight=None, fit=True, ax=None):
...
-# control this behaviour. A small value of learning-rate will only correct the
-# residuals of very few samples. If the learning-rate is set to 1 then we will
-# find a tree that fit the residuals of all samples. So, with a very low
+# Finally, we have overlooked the impact of the `learning_rate` parameter up
+# till now. When
+# fitting the residuals one could choose if the tree should try to correct all
+# possible errors or only a fraction of them. The learning-rate allows you to
+# control this behaviour. A small learning-rate value would only correct the
+# residuals of very few samples. If a large learning-rate is set (e.g., 1),
+# we would fit the residuals of all samples. So, with a very low
```
Comment (lucyleeow) [inline python_scripts/ensemble.py:947]: I would appreciate more detail here - how does changing the learning-rate affect how many residuals are corrected? But maybe it's too much detail / outside the scope of this MOOC.

File: python_scripts/ensemble.py [hunk 6]
```diff
@@ -36,13 +37,14 @@
...
+# `min_samples_leaf`.
 #
-# We recall that we need to tune these parameters because the decision trees
-# tend to overfit the training data if we grow deep the trees but there are no
-# rules to limit the parameters. Thus, not making a search could lead us to
-# have an underfitted model.
+# We recall that we need to tune these parameters, as decision trees
+# tend to overfit the training data if we grow deep trees, but there are no
+# rules on how to limit the parameters. Thus, not making a search could lead us
+# to have an underfitted model.
```
[the remark this answers is not in the document]
Comment (lucyleeow) [reply python_scripts/ensemble.py:47]: Fix in https://github.com/lucyleeow/scikit-learn-mooc/pull/1/files

File: python_scripts/ensemble.py [hunk 7]
```diff
@@ -117,15 +120,15 @@
 
 # %% [markdown]
 # Hence, we have a model that has an $R^2$ score below 0.7. The amount of time
-# to find the best learner depends on the number of fold used during the
+# to find the best learner depends on the number of folds used during the
 # cross-validation in the grid-search multiplied by the number of parameters.
 # Therefore, the computational cost is quite high.
```
[the remark this answers is not in the document]
Comment (lucyleeow) [reply python_scripts/ensemble.py:125]: No problem for repeating. My point is more that we should include this information in the 'basic parameter tuning' notebook as well. Here we say:
https://github.com/INRIA/scikit-learn-mooc/blob/76bbf56bbf4ca98ffc5742e1efffe5a57f2775f4/python_scripts/04_basic_parameters_tuning.py#L215-L218

But we can expand this in 'basic tuning' and add the information about fold as well.

File: python_scripts/ensemble.py [hunk 8]
```diff
@@ -148,23 +151,24 @@
...
 # baseline with decent performance without any parameter tuning.
 #
-# Now, we will go into details by presenting two ensemble families: bagging and
+# Now, we will discuss in detail two ensemble families: bagging and
 # boosting.
 #
 # ## Bagging
 #
-# Bagging stands for bootstrap aggregating. Indeed, it uses bootstrap samples
+# Bagging stands for Bootstrap AGGregatING. It uses bootstrap samples
```
[the remark this answers is not in the document]
Comment (lucyleeow) [reply python_scripts/ensemble.py:166]: Fix in https://github.com/lucyleeow/scikit-learn-mooc/pull/1/files

File: python_scripts/ensemble.py [hunk 9]
```diff
@@ -254,7 +258,7 @@ def bootstrap_sample(x, y):
 
 # %% [markdown]
 # We observe that the 3 generated bootstrap samples are all different. To
-# confirm this intuition, we can check the quantity of unique sample in the
+# confirm this intuition, we can check the number of unique samples in the
 # bootstrap samples.
 
 # %%
```
[the remark this answers is not in the document]
Comment (lucyleeow) [reply python_scripts/ensemble.py:264]: Probably easier to chat offline. I'll note this one down.

File: python_scripts/ensemble.py [hunk 10]
```diff
@@ -545,20 +552,20 @@ def plot_decision_function(X, y, clf, sample_weight=None, fit=True, ax=None):
 ax.legend()
```
[the remark this answers is not in the document]
Comment (lucyleeow) [reply python_scripts/ensemble.py:553]: Fixed in https://github.com/lucyleeow/scikit-learn-mooc/pull/1/

File: python_scripts/ensemble.py [hunk 11]
```diff
@@ -577,7 +584,8 @@ def plot_decision_function(X, y, clf, sample_weight=None, fit=True, ax=None):
```
[the remark this answers is not in the document]
Comment (lucyleeow) [reply python_scripts/ensemble.py:585]: fixed in https://github.com/lucyleeow/scikit-learn-mooc/pull/1/

File: python_scripts/ensemble.py [hunk 12]
```diff
@@ -607,29 +615,28 @@ def plot_decision_function(X, y, clf, sample_weight=None, fit=True, ax=None):
...
 #   to samples;
-# * one needs to assign a weight to each learner when making the predictions.
+# * one needs to assign a weight to each learner when making predictions.
 #
 # Indeed, we defined a really simple scheme to assign sample weights and
-# learner weights. However, there are some statistical basis which
-# theoretically define how these weights should be assigned.
+# learner weights. However, there are statistical theory for how these
+# these sample and learner weights can be optimally calculated.
 # FIXME: I think we should add a reference to ESL here.
```
[the remark this answers is not in the document]
Comment (lucyleeow) [reply python_scripts/ensemble.py:637]: see https://github.com/lucyleeow/scikit-learn-mooc/pull/1/

File: python_scripts/ensemble.py (hunk 1 above)
Comment (lucyleeow) [reply python_scripts/ensemble.py:847]: see https://github.com/lucyleeow/scikit-learn-mooc/pull/1

File: python_scripts/ensemble.py [hunk 13]
```diff
@@ -651,28 +658,30 @@ def plot_decision_function(X, y, clf, sample_weight=None, fit=True, ax=None):
...
-# decision tree (GBDT) algorithm which we will present now.
+# decision tree (GBDT) algorithm which we will discuss now.
 #
 # ### Gradient-boosting decision tree (GBDT)
 #
 # Gradient-boosting differs from AdaBoost due to the following reason: instead
-# of assigning a weights to some specific samples, GBDT will fit a decision
-# tree on the residuals (hence the name "gradient") from the previous trees.
+# of assigning weights to specific samples, GBDT will fit a decision
+# tree on the residuals (hence the name "gradient") of the previous tree.
```
[the remark this answers is not in the document]
Comment (TwsThomas) [reply python_scripts/ensemble.py:678]: Okay, for me also the implication of the name "gradient" is a bit fast here

File: python_scripts/ensemble.py (hunk 2 above)
Comment (TwsThomas) [reply python_scripts/ensemble.py:928]: My understanding is that in gradient-boosting, hyper parameters are strongly coupled, just as `max_depth` and `learning_rate`. 
But I guess that hyper parameters are usually always a bit coupled...
So, in my opinion
> instead of setting one after the other each parameter

seems to be a bit inappropriate, since we always use grid search.

File: python_scripts/ensemble.py [hunk 14]
```diff
@@ -410,18 +413,19 @@ def bootstrap_sample(x, y):
...
+# it to have the same behaviour as the random forest classifier.
 #
-# ### Specificities regarding the classifiers.
+# ### Classifiers details
 #
-# Up to now, we only focused on regression problems. There is little a
+# Up to now, we have only focused on regression problems. There is a little
 # difference between regression and classification.
 #
 # First, the `base_estimator` should be chosen in line with the problem that
```
[the remark this answers is not in the document]
Comment (lesteve) [reply python_scripts/ensemble.py:432]: I think it is fine, better to err on the cautious side of obvious ;-)

File: python_scripts/ensemble.py (hunk 3 above)
Comment (lesteve) [reply python_scripts/ensemble.py:940]: Good enough for now.

File: python_scripts/ensemble.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lesteve) [reply python_scripts/ensemble.py:553]: Not sure what the original strategy was, but can you open a PR on this repo with your changes?

File: python_scripts/ensemble.py (hunk 5 above)
Comment (lesteve) [reply python_scripts/ensemble.py:948]: Probably too much detail.

**Comment by lesteve** [comment]
I am going to merge this one, this seems like a great improvement!

@lucyleeow about the points that were not resolved in this PR, feel free to open separate PRs/issues!

**Comment by lucyleeow** [comment]
Meep @TwsThomas actually opened a PR on my fork to address some of the issues: https://github.com/lucyleeow/scikit-learn-mooc/pull/1 We were working on merging his PR before merging this one. I guess he could change his PR on my fork to a PR in this repo ?

**Comment by lesteve** [comment]
> I guess he could change his PR on my fork to a PR in this repo ?

Probably. In principe @TwsThomas and you (@lucyleeow) should have the necessary rights to push into a contributor PR branch, which is probably the easiest way of collaborating in a single PR.

**Comment by lucyleeow** [comment]
Oh yeah, we could have done it that way too.
